atomic_reactor/config.py

```diff
@@ -0,0 +1,568 @@
+        except KeyError:
+            if fallback != NO_FALLBACK:
+                return fallback
+            raise
+
+    @property
+    def koji(self):
+        koji_map = self._get_value(ReactorConfigKeys.KOJI_KEY)
+
+        if 'auth' in koji_map:
+            krb_principal = koji_map['auth'].get('krb_principal')
+            krb_keytab = koji_map['auth'].get('krb_keytab_path')
+            if bool(krb_principal) != bool(krb_keytab):
+                raise RuntimeError("specify both koji_principal and koji_keytab or neither")
+
+        return koji_map
+
+    @property
+    def koji_session(self):
+        from atomic_reactor.utils.koji import create_koji_session
+
+        auth_info = {
+            "proxyuser": self.koji['auth'].get('proxyuser'),
+            "ssl_certs_dir": self.koji['auth'].get('ssl_certs_dir'),
+            "krb_principal": self.koji['auth'].get('krb_principal'),
+            "krb_keytab": self.koji['auth'].get('krb_keytab_path')
+        }
+
+        use_fast_upload = self.koji.get('use_fast_upload', True)
+
+        return create_koji_session(self.koji['hub_url'], auth_info, use_fast_upload)
+
+    @property
+    def koji_path_info(self):
+        from koji import PathInfo
+
+        # Make sure koji root_url doesn't end with a slash since the url
+        # is used to construct resource urls (e.g. log links)
+        root_url = self.koji['root_url'].rstrip('/')
+        return PathInfo(topdir=root_url)
+
+    @property
+    def pnc(self):
+        return self._get_value(ReactorConfigKeys.PNC_KEY, fallback={})
+
+    @property
+    def odcs(self):
+        return self._get_value(ReactorConfigKeys.ODCS_KEY, fallback={})
+
+    @property
+    def odcs_session(self):
+        from atomic_reactor.utils.odcs import ODCSClient
+
+        client_kwargs = {
+            'insecure': self.odcs.get('insecure', False),
+            'timeout': self.odcs.get('timeout', None),
+        }
+
+        openidc_dir = self.odcs['auth'].get('openidc_dir')
+        if openidc_dir:
+            token_path = os.path.join(openidc_dir, 'token')
+            with open(token_path, "r") as f:
+                client_kwargs['token'] = f.read().strip()
+
+        ssl_certs_dir = self.odcs['auth'].get('ssl_certs_dir')
+        if ssl_certs_dir:
+            cert_path = os.path.join(ssl_certs_dir, 'cert')
+            if os.path.exists(cert_path):
+                client_kwargs['cert'] = cert_path
+            else:
+                raise KeyError("ODCS ssl_certs_dir doesn't exist")
+
+        return ODCSClient(self.odcs['api_url'], **client_kwargs)
+
+    @property
+    def smtp(self):
+        return self._get_value(ReactorConfigKeys.SMTP_KEY, fallback={})
+
+    @property
+    def smtp_session(self):
+        import smtplib
+        return smtplib.SMTP(self.smtp['host'])
+
+    @property
+    def cachito(self):
+        return self._get_value(ReactorConfigKeys.CACHITO_KEY, fallback={})
+
+    @property
+    def cachito_session(self):
```

Comment:
`*_session` methods shouldn't be part of class, you can keep them in module but it's very weird having session property as part of configuration

Reply:
*_session moved outside of class as standalone functions